File: works/tmp/2512.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.Collections;



public class Main
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] arr = new int[10001];

        int N = Integer.parseInt(br.readLine());
        int maxVal = 0;
        String[] _s = br.readLine().split(" ");
        for(int i = 0; i < N; i++)
        {
            arr[i] = Integer.parseInt(_s[i]);
            if(arr[i] > maxVal) maxVal = arr[i];
        }
        int M = Integer.parseInt(br.readLine());

        int left = 0, right = maxVal, answer = 0;
        while(left <= right)
        {
            int mid = (left + right) / 2;
            int sum = 0;
            for(int i = 0; i < N; i++)
            {
                if(arr[i] > mid) sum += mid;
                else sum += arr[i];
            }

            if(sum <= M)
            {
                left = mid + 1;
                answer = mid;
            }
            else
            {
                right = mid - 1;
            }
        }
        System.out.print(answer);
    }
}

File: works/2512.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*


fun main()
{
    val br = BufferedReader(InputStreamReader(System.`in`))
    var arr = IntArray(10001)
    var N = br.readLine().toInt()
    var maxVal = 0
    val _s = br.readLine().split(' ')
    for(i in 0 until N)
    {
        arr[i] = _s[i].toInt()
        if(arr[i] > maxVal) maxVal = arr[i]
    }
    var M = br.readLine().toInt()

    var left = 0; var right = maxVal; var answer = 0
    while(left <= right)
    {
        var mid = (left + right) / 2
        var sum = 0
        for(i in 0 until N)
        {
            if(arr[i] > mid) sum += mid
            else sum += arr[i]
        }
        if(sum <= M)
        {
            left = mid + 1
            answer = mid
        }
        else
        {
            right = mid - 1
        }
    }
    print(answer)
}